---
title: 消除异步的传染性：简单理解和解决方法
pubDatetime: 2024-08-31
draft: true
tags:
  - JavaScript
---

## 前言

在 JavaScript 中，我们经常需要处理异步操作，比如通过 `fetch` 发送网络请求。但是，这些异步操作有一个特点，那就是它们会“传染”——如果一个函数是异步的，那么所有调用它的函数也必须是异步的。这种情况可能会让整个程序变得复杂。

```ts
async function getUser() {
  return await fetch("./1.json");
}

async function m1() {
  const user = await getUser();
  return user;
}

async function m2() {
  const user = await m1();
  return user;
}

async function m3() {
  const user = await m2();
  return user;
}

async function main() {
  const user = await m3();
  console.log(user);
}
```

在这个例子中，最开始的 `getUser` 函数是异步的，因为它使用了 `fetch`，导致 `m1`、`m2`、`m3` 以及 `main` 也都成了异步函数。 这就像多米诺骨牌一样，一倒全倒。这在一些场景中可能会增加代码的复杂度，也违反了函数式编程中关于“纯函数”的原则。

## 原因分析

问题的根源在于 `getUser` 函数，它使用了 `fetch` 发送网络请求。网络请求是异步的，意味着我们不能立刻得到结果。为了避免阻塞浏览器的主线程，JavaScript 选择在后台处理请求，而这段时间里继续执行其他代码。

但这也意味着任何依赖 `getUser` 结果的函数必须等待它完成，进而变成了异步函数。

![](https://s2.loli.net/2024/08/28/mC1MtPA9pDLEkiy.png)

如图所示，这是函数的调用时间线。由于 `fetch` 需要等待，那么 `getUser` 也需要等待，从而 `mian` 函数也需要等。那么最终只能在 `fetch` 做文章。

## 解决方案

我们的目标是让这些函数看起来像同步的，但又不能真的阻塞浏览器主线程。一个方法是利用缓存机制：在第一次调用 `fetch` 时，立即发出请求，但不等待结果，而是抛出一个错误。下一次调用时，如果有缓存，就立刻返回缓存结果。

![](https://s2.loli.net/2024/08/28/tAZXdEjNwBil8zn.png)

在这种实现中，`fetch` 函数不会等待请求完成，而是立即抛出一个错误。尽管 `main` 函数因此出错并中断执行，但网络请求已经发出，不会被取消。当请求成功后，我们可以将结果缓存起来（即上一次请求的结果），然后重新开始整个调用链，从头再执行一次 `main` 函数（注意，这次执行的 `main` 已经不是第一次的 `main`，而是重新运行后的版本）。

重新执行 `main` 函数后，它会再次调用 `fetch`。这一次，因为结果已经缓存，`fetch` 可以立即返回结果而无需等待。这样一来，函数调用链就变成了同步操作。虽然整个过程需要经历两次调用，但第一次的错误处理保证了最终的同步效果，第二次调用可以顺利结束而不再等待异步操作。

### 重写 Fetch 逻辑

![](https://s2.loli.net/2024/08/28/JkaPb34xKhVG68d.png)

当调用 `fetch` 请求函数时，首先检查是否有缓存的结果。如果有缓存，直接返回缓存中的结果，这样就不需要等待新的请求了。如果没有缓存，就立即发送请求，并用 `Promise` 将请求的结果写入缓存，但同时不等待这个请求的完成，而是立刻抛出一个错误。

在这种机制下，下次调用 `fetch` 时，如果缓存中已经有结果，就可以直接返回缓存的结果，实现同步操作的效果。

```ts
function run(func) {
  let cache = {
    status: "pending", // 状态：pending, fulfilled, rejected
    value: null,
  };
  const oldFetch = window.fetch;

  window.fetch = function (...args) {
    if (cache.status === "fulfilled") {
      // 如果有缓存，直接返回缓存的结果
      return cache.value;
    } else if (cache.status === "rejected") {
      // 如果有缓存，但请求失败，抛出错误
      throw cache.value;
    }

    // 第一次请求，发出请求并缓存结果
    const _promise = oldFetch(...args).then(
      res => {
        cache.status = "fulfilled";
        cache.value = res;
      },
      error => {
        cache.status = "rejected";
        cache.value = error;
      }
    );

    // 立即抛出错误
    throw _promise;
  };

  // 处理第一次抛出的错误，重试请求
  try {
    func();
  } catch (error) {
    if (error instanceof Promise) {
      error.then(func, func).finally(() => {
        window.fetch = oldFetch; // 恢复原始 fetch 方法
      });
    }
  }
}

run(main);
```

## 参考资料

- [用动图的方式了解 JavaScript 异步（一）]()
- [用动图的方式了解 JavaScript 异步（二）]()
